internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,139 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the AccessToken method.
+type Tokener interface {
+	Token(context.Context) (string, error)
+	ForceRefresh(context.Context) (string, error)
+}
+
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// 	AccessToken string holds the currently cached token.
+	AccessToken string
+	// 	RefreshToken string token after refresh.
+	RefreshToken string
+	// Expiration stores the expiration time of the current token.
+	Expiration time.Time
+	// RefreshExpiration stores the expiration time of the refresh token.
+	RefreshExpiration time.Time
+	// TokenType is the type of the access token (e.g., "Bearer").
+	TokenType string
+	// NotBeforePolicy is the not-before policy of the access token.
+	NotBeforePolicy int
+	// Scope is the scope granted for the access token.
+	Scope string
+	// mutex ensures safe concurrent access to the token and expiration fields.
+	mutex sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken      string `json:"access_token"`
+	ExpiresIn        int    `json:"expires_in"`
+	RefreshExpiresIn int    `json:"refresh_expires_in"`
```

Comment:
Did you manage to get a refresh token out of the stage/prod services? I could not.

Reply:
we can remove the RefreshExpiresIn, as you mentioned 'it appears that refresh token is actually optional, our service does not appear to return it. In that case, we can stick with the initial goal - when access token expires, just request a new one with id/secret.' 
no?